### PR TITLE
fix(app): stabilize session view state

### PR DESCRIPTION
**Pull request by Astro-Han**
## Summary

- Add a session view controller that separates route session state from visible session state.
- Route timeline, composer state, prompt submit, abort, handoff, permission, question, and todo surfaces through the visible session while route data is still loading.
- Strengthen session scroll and switch E2E coverage so regressions catch top jumps, partial render windows, mixed-session frames, composer unmounts, and page errors.

## Why

Fixes the recurring session UI class behind #286, #312, and #313: URL route state could move ahead of the fully rendered session, while timeline and composer code read mixed sources. That could cause old-hash submit scroll jumps, session switch flicker, or composer actions targeting the wrong session during transition.

## Related Issue

Related to #286, #312, and #313.

## How To Verify

```bash
bun test --preload ./happydom.ts src/components/prompt-input/submit.test.ts src/pages/session/session-view-controller.test.ts src/pages/session/timeline-session-state.test.ts
bun run --cwd packages/app typecheck
git diff --check
bun run --cwd packages/app test:e2e -- session/session-scroll-position.spec.ts
```

## Screenshots or Recordings

No screenshot attached. This is a visible session-transition behavior fix covered by the targeted Playwright E2E checks listed above.

## Checklist

- [x] I linked the related issue, or stated why there is no issue
- [x] This PR has type, scope, and priority labels, or I requested maintainer labeling
- [x] I listed the relevant verification steps, including tests when behavior changed
- [x] I manually checked visible UI or copy changes when needed, with screenshots or recordings
- [x] I considered macOS and Windows impact for desktop, packaging, updater, signing, paths, shell, or permissions changes
- [x] I called out docs, release notes, dependencies, permissions, credentials, deletion behavior, or generated/local file changes when relevant
- [x] I am targeting `dev`, and my PR title and commit messages use Conventional Commits in English

## Summary by CodeRabbit

* **New Features**
  * Composer and prompt input can be driven by an externally provided session identity; timeline view separates route vs visible session state for smoother session rendering.

* **Tests**
  * New and expanded e2e and unit tests covering session view transitions, scroll-position preservation, submit-via-modifier behavior, and page-error capture.

* **Bug Fixes**
  * Improved session consistency during submissions, session switches, restores, and related UI transitions.

### CODE REVIEW

**Comment by Astro-Han** [comment]
## Design Review: Session View Controller

Overall design is solid — decoupling route session from visible session via `SessionViewController` cleanly fixes the mixed-source transition bugs.

**No blocking (P0/P1) issues found.**

| Severity | Count | Highlights |
|----------|-------|------------|
| P0 | 0 | — |
| P1 | 0 | — |
| P2 | 2 | Dead code, naming inconsistency |
| P3 | 2 | Duplicate tests, fragile comparator |

### Edge cases verified ✅

- Navigate to new session, messages not ready → visible stays on old session ✅
- Messages load → swap to new session ✅
- Navigate away from session entirely → visible cleared ✅
- Directory changes mid-transition → stale state cleared ✅
- Composer/submit/abort target visible session during transition ✅
- `deferRender` keyed on `timelineSessionKey` (correct) ✅

### Regression risk: Low

The PR is additive. Callers that don't pass `sessionID`/`isNewSession` fall back to `params.id` (old behavior). Home page composer is unaffected.

One area to watch: the `deferRender` key change from `sessionKey()` → `timelineSessionKey()` shifts when the composer flash happens. The E2E session-switch test covers this.

> Inline comments posted below on specific files/lines.

File: packages/app/src/pages/session/session-view-controller.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  directory: Accessor<string>
+  routeSessionID: Accessor<string | undefined>
+  routeMessagesReady: Accessor<boolean>
+}
+
+export function sessionKey(input: { directory: string; sessionID: string | undefined }) {
+  return `${input.directory}${input.sessionID ? `/${input.sessionID}` : ""}`
+}
+
+export function nextTimelineSessionID(input: {
```
Comment (Astro-Han) [inline packages/app/src/pages/session/session-view-controller.ts]: **P2 — Naming: `nextTimelineSessionID` is now misleading.**

This function no longer controls just the timeline — it drives the entire visible session (composer, todos, submit target). Consider renaming to `nextVisibleSessionID` and keeping the old name as a re-export alias in `timeline-session-state.ts` for backward compat.

File: packages/app/src/pages/session/composer/session-composer-state.ts [hunk 2]
```diff
@@ -18,26 +18,27 @@ const todoTerminal = (todo: Todo) => todo.status === "completed" || todo.status
 
 const todoSignature = (todos: Todo[]) => todos.map((todo) => `${todo.status}:${todo.content}`).join("\u0000")
 
-export function createSessionComposerState() {
+export function createSessionComposerState(input?: { sessionID: () => string | undefined }) {
   const params = useParams()
```
Comment (Astro-Han) [inline packages/app/src/pages/session/composer/session-composer-state.ts]: **P2 — Dead code: `params` is no longer used after this refactor.**

`useParams()` is still called but all references to `params.id` have been replaced by `activeSessionID()`. The `params` variable is now dead code — remove it to avoid confusion about which source of truth is active.

File: packages/app/src/pages/session/timeline-session-state.test.ts [hunk 3]
```diff
@@ -32,3 +32,69 @@ describe("nextTimelineSessionID", () => {
...
+
+describe("sessionKey", () => {
+  test("keys a directory route without a session id", () => {
+    expect(sessionKey({ directory: "repo", sessionID: undefined })).toBe("repo")
+  })
+
+  test("keys a concrete session route", () => {
+    expect(sessionKey({ directory: "repo", sessionID: "ses_target" })).toBe("repo/ses_target")
+  })
+})
```
Comment (Astro-Han) [inline packages/app/src/pages/session/timeline-session-state.test.ts]: **P2 — Duplicate test coverage.**

`nextSessionViewState` is tested here and in `session-view-controller.test.ts` with near-identical cases (loading, ready, clear on leave). Recommend keeping the canonical tests in `session-view-controller.test.ts` and reducing this file to only test the re-export wiring, to avoid maintenance drift.

File: packages/app/e2e/session/session-scroll-position.spec.ts [hunk 4]
```diff
@@ -159,59 +244,126 @@ async function seedSessionTurns(input: { sdk: Sdk; sessionID: string; count: num
...
+      turnListSelector,
+    }) => {
+      const routeSessionID = () => window.location.pathname.match(/\/session\/([^/?#]+)/)?.[1]
+      const removedMatches = (node: Node, selector: string) =>
+        node instanceof Element && (node.matches(selector) || !!node.querySelector(selector))
+      const read = (removed?: { composerDock?: boolean; messageList?: boolean }) => {
+        const list = document.querySelector(turnListSelector)
+        const viewport = list?.closest(scrollViewportSelector)
+        const timeline = viewport instanceof HTMLElement ? viewport : undefined
+        const composer = document.querySelector(composerDockSelector)
```
Comment (Astro-Han) [inline packages/app/e2e/session/session-scroll-position.spec.ts:266]: **P3 — Fragile array comparison.**

`.join(",")` works but would silently produce incorrect dedup results if a session ID ever contains a comma. Consider `JSON.stringify` or element-wise comparison for robustness.

File: packages/app/src/pages/session/timeline-session-state.test.ts (hunk 3 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/timeline-session-state.test.ts]: P2 fixed in bd78760b5. timeline-session-state.test.ts now only verifies legacy re-export wiring; canonical behavior remains in session-view-controller.test.ts.

File: packages/app/src/pages/session/session-view-controller.ts (hunk 1 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/session-view-controller.ts]: P2 fixed in bd78760b5. Renamed the canonical helper to nextVisibleSessionID and kept nextTimelineSessionID as the legacy re-export from timeline-session-state.ts.

File: packages/app/e2e/session/session-scroll-position.spec.ts (hunk 4 above)
Comment (Astro-Han) [reply packages/app/e2e/session/session-scroll-position.spec.ts:266]: P3 fixed in bd6ff8dec. Replaced comma-join comparison with elementwise list comparison.

File: packages/app/src/pages/session/session-view-controller.test.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      expect(controller.route.ready()).toBe(false)
+      expect(controller.visible.id()).toBeUndefined()
+      expect(controller.visible.key()).toBe("repo")
+      expect(controller.visible.ready()).toBe(false)
+      expect(controller.transitioning()).toBe(true)
+
+      dispose()
+    })
+  })
+})
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/app/src/pages/session/session-view-controller.test.ts:45]: P3, intentionally left unchanged. I verified the current Bun plus Solid test environment does not exercise createMemo signal recomputation the way the browser runtime does, so this specific signal-driven test would be misleading. The pure transition rules remain covered by nextSessionViewState, and the browser path is covered by the session switch E2E.

File: packages/app/src/pages/session/composer/session-composer-state.ts (hunk 2 above)
Comment (Astro-Han) [reply packages/app/src/pages/session/composer/session-composer-state.ts]: P2 fixed in bd78760b5. createSessionComposerState now requires the caller-provided sessionID accessor and no longer reads useParams.

File: packages/app/e2e/session/session-scroll-position.spec.ts [hunk 6]
```diff
@@ -135,13 +159,74 @@ async function stopTimelineScrollProbe(page: Page) {
...
+  }
+  const handler = (error: Error) => {
+    errors.push({ type: "pageerror", message: describe(error) })
+  }
+  page.on("pageerror", handler)
+  return {
+    errors,
+    dispose: () => page.off("pageerror", handler),
+  }
 }
```
[the remark this answers is not in the document]
Comment (Astro-Han) [reply packages/app/e2e/session/session-scroll-position.spec.ts:230]: P3, intentionally left unchanged. The two formatters live in different execution contexts: one is serialized into page.addInitScript and one runs in the Node side Playwright listener. Extracting a shared helper would add cross-context indirection for a small diagnostic formatter, so YAGNI here.

**Comment by Astro-Han** [comment]
Handled the two non-inline CodeRabbit findings in 78d0c4cc2.

- P1: followup and revert composer gating now use the visible session identity instead of the route session.
- P2: todo refresh now tracks the visible session identity and drops stale refreshes when visible state changes.

Verification after review fixes:
- bun run --cwd packages/app typecheck
- bun --cwd packages/app test -- src/pages/session/session-view-controller.test.ts src/pages/session/timeline-session-state.test.ts src/pages/session/composer/session-composer-state.test.ts
- bun run --cwd packages/app test:e2e -- session/session-scroll-position.spec.ts

**Comment by Astro-Han** [comment]
Handled the two latest outside-diff comments in 8d25fc076.

- P2: artifact history and Files-derived artifacts now key off the visible session via timelineSessionID(), and stale artifact results are compared against that same visible id.
- P2: last-turn state now derives from timelineVisibleUserMessages(), so turn diffs, turn-mode review, and syncSessionModel follow the same visible session as timeline and composer during route transitions.

Verification before pushing:
- bun run --cwd packages/app typecheck
- bun --cwd packages/app test -- src/pages/session/session-view-controller.test.ts src/pages/session/timeline-session-state.test.ts src/pages/session/session-model-helpers.test.ts
- bun run --cwd packages/app test:e2e -- session/session-scroll-position.spec.ts
